designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,338 @@
+# Local Debugging Experience for SAM Applications
+
+Current Status: Not Implemented
+
+## Introduction
+
+The AWS Toolkit enhances the SAM Application development experience by integrating local debug functionality into VS Code. This document outlines the available functionality.
+
+While this document's main focus is on debugging capabilities in the toolkit, there are places where the experience around invoking without the debugger (aka "running") is also discussed.
+
+Each programming language (and corresponding Lambda Runtimes) requires Toolkit support for debugging features to work. A limited selection of programming languages are supported in the Toolkit.
+
+### Terminology
+
+#### CodeLens
+
+CodeLenses are visual decorators anchored to document locations. They are used to convey information and/or provide links that trigger an action. Additional information and examples about CodeLenses can be found [on the VS Code blog](https://code.visualstudio.com/blogs/2017/02/12/code-lens-roundup).
+
+#### Debug Configuration
+
+Debug Configurations are user-managed JSON entries that define what programs can be debugged. Pressing F5 (or the Debug button) starts a Debug session for the Debug Configuration currently selected in VS Code's Debug View. VS Code extensions increase VS Code's debugging capablities by implementing Debug Configuration types.
+
+Debug Configurations are stored in `.vscode/launch.json` relative to the VS Code workspace.
+
+More information about VS Code Debugging can be found [in the VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).
+
+#### SAM Template
+
+A SAM Template defines a Serverless Application's resources, and supporting code. This is used by the SAM CLI to build, run, package, and deploy the Application.
+
+Additional information about SAM can be found at:
+
+-   [SAM Homepage](https://aws.amazon.com/serverless/sam/)
+-   [What Is the AWS Serverless Application Model (AWS SAM)?](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html)
+-   [SAM CLI GitHub Repo](https://github.com/awslabs/aws-sam-cli)
+
+## Overview
+
+The toolkit supports the following scenarios for Locally Running and Debugging code using the Serverless Application Model:
+
+-   invoking SAM Template resources that are Lambda functions
+-   making API Gateway style requests against SAM Template resources that are Lambda functions
+-   invoking standalone Lambda function handlers (these don't use SAM Templates, but the debugging functionality is supported by one behind the scenes)
```

Comment:
Can we add a supplemental experience where we make the generated SAM template explicit? i.e. we allow the user to save this to their project.

This would allow the user to use the SAM CLI as well. It would also let the user explore and customize their SAM settings in the sam template if they continue to use SAM.